{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "# WGCNA Analysis for Breast Cancer Subtype Classification\n",
        "\n",
        "This script performs Weighted Gene Co-expression Network Analysis (WGCNA)\n",
        "to identify gene modules associated with breast cancer subtypes"
      ],
      "metadata": {
        "id": "hTWYe1DG2i-H"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-IwUFCy1o_uQ",
        "outputId": "d6b2c1bc-1a55-4186-beb3-cffa77311e5e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.22 (BiocManager 1.30.27), R 4.5.2 (2025-10-31)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'DESeq2'\n",
            "\n",
            "also installing the dependencies ‘XVector’, ‘formatR’, ‘abind’, ‘SparseArray’, ‘lambda.r’, ‘futile.options’, ‘Seqinfo’, ‘S4Arrays’, ‘DelayedArray’, ‘futile.logger’, ‘snow’, ‘BH’, ‘S4Vectors’, ‘IRanges’, ‘GenomicRanges’, ‘SummarizedExperiment’, ‘BiocGenerics’, ‘Biobase’, ‘BiocParallel’, ‘matrixStats’, ‘locfit’, ‘MatrixGenerics’, ‘RcppArmadillo’\n",
            "\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.22 (BiocManager 1.30.27), R 4.5.2 (2025-10-31)\n",
            "\n",
            "Installing package(s) 'org.Hs.eg.db'\n",
            "\n",
            "also installing the dependencies ‘plogr’, ‘png’, ‘Biostrings’, ‘RSQLite’, ‘KEGGREST’, ‘AnnotationDbi’\n",
            "\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.22 (BiocManager 1.30.27), R 4.5.2 (2025-10-31)\n",
            "\n",
            "Installing package(s) 'impute'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.22 (BiocManager 1.30.27), R 4.5.2 (2025-10-31)\n",
            "\n",
            "Installing package(s) 'preprocessCore'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.22 (BiocManager 1.30.27), R 4.5.2 (2025-10-31)\n",
            "\n",
            "Installing package(s) 'GO.db'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.22 (BiocManager 1.30.27), R 4.5.2 (2025-10-31)\n",
            "\n",
            "Installing package(s) 'WGCNA'\n",
            "\n",
            "also installing the dependencies ‘checkmate’, ‘gridExtra’, ‘htmlTable’, ‘colorspace’, ‘Formula’, ‘iterators’, ‘dynamicTreeCut’, ‘fastcluster’, ‘Hmisc’, ‘foreach’, ‘doParallel’\n",
            "\n",
            "\n",
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "also installing the dependency ‘plyr’\n",
            "\n",
            "\n",
            "Loading required package: dynamicTreeCut\n",
            "\n",
            "Loading required package: fastcluster\n",
            "\n",
            "\n",
            "Attaching package: ‘fastcluster’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:stats’:\n",
            "\n",
            "    hclust\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘WGCNA’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:stats’:\n",
            "\n",
            "    cor\n",
            "\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "Loading required package: generics\n",
            "\n",
            "\n",
            "Attaching package: ‘generics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,\n",
            "    setequal, union\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,\n",
            "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
            "    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,\n",
            "    unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:utils’:\n",
            "\n",
            "    findMatches\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: IRanges\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: Seqinfo\n",
            "\n",
            "Loading required package: SummarizedExperiment\n",
            "\n",
            "Loading required package: MatrixGenerics\n",
            "\n",
            "Loading required package: matrixStats\n",
            "\n",
            "\n",
            "Attaching package: ‘MatrixGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
            "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
            "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
            "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
            "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
            "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
            "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
            "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
            "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
            "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
            "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
            "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
            "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
            "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
            "    rowWeightedSds, rowWeightedVars\n",
            "\n",
            "\n",
            "Loading required package: Biobase\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘Biobase’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:MatrixGenerics’:\n",
            "\n",
            "    rowMedians\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    anyMissing, rowMedians\n",
            "\n",
            "\n",
            "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
            "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.6\n",
            "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.1     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.6.0\n",
            "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 4.0.1     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.3.1\n",
            "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.4     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.2\n",
            "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.2.1     \n",
            "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
            "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mIRanges\u001b[39m::collapse()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m        masks \u001b[34mmatrixStats\u001b[39m::count()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::desc()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mstats\u001b[39m::filter()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::first()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mS4Vectors\u001b[39m::second()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mIRanges\u001b[39m::slice()\n",
            "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
            "\n",
            "Attaching package: ‘reshape2’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:tidyr’:\n",
            "\n",
            "    smiths\n",
            "\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Allowing parallel execution with up to 2 working processes.\n"
          ]
        }
      ],
      "source": [
        "# If needed (run once)\n",
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE)) install.packages(\"BiocManager\")\n",
        "\n",
        "# Bioconductor packages and WGCNA (as BiocManager handles its dependencies, explicitly listing WGCNA's Bioc dependencies for robustness)\n",
        "pkgs_bioc_and_wgcna <- c(\"DESeq2\", \"org.Hs.eg.db\", \"impute\", \"preprocessCore\", \"GO.db\", \"AnnotationDbi\", \"WGCNA\")\n",
        "for (p in pkgs_bioc_and_wgcna) {\n",
        "    if (!requireNamespace(p, quietly = TRUE)) {\n",
        "        BiocManager::install(p, update = FALSE, ask = FALSE)\n",
        "    }\n",
        "}\n",
        "\n",
        "# Other CRAN packages\n",
        "pkgs_cran_others <- c(\"tidyverse\", \"pheatmap\", \"reshape2\", \"ggplot2\") # Removed flashClust\n",
        "for (p in pkgs_cran_others) {\n",
        "    if (!requireNamespace(p, quietly = TRUE)) {\n",
        "        install.packages(p)\n",
        "    }\n",
        "}\n",
        "\n",
        "library(WGCNA)\n",
        "library(DESeq2)\n",
        "library(tidyverse)\n",
        "library(pheatmap)\n",
        "# library(flashClust) # Removed flashClust\n",
        "library(reshape2)\n",
        "library(ggplot2)\n",
        "\n",
        "# Important for WGCNA\n",
        "options(stringsAsFactors = FALSE)\n",
        "enableWGCNAThreads()"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "###  Loading and Preparing Data"
      ],
      "metadata": {
        "id": "BjyU-lfDToit"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Load normalized count data\n",
        "count_data <- read.csv(\"filtered_count_data.csv\", row.names = 1)\n",
        "\n",
        "\n",
        "# Load sample metadata\n",
        "col_data <- read.csv(\"col_data.csv\")\n",
        "# Ensure Sample column matches count_data column names\n",
        "rownames(col_data) <- col_data$Sample\n",
        "\n",
        "# Transpose the data: WGCNA expects samples as rows, genes as columns\n",
        "datExpr <- t(count_data)\n",
        "\n",
        "\n",
        "# Checking for genes and samples with too many missing values\n",
        "gsg <- goodSamplesGenes(datExpr, verbose = 3)\n",
        "if (!gsg$allOK) {\n",
        "    # Removing offending genes and samples\n",
        "    if (sum(!gsg$goodGenes) > 0)\n",
        "        printFlush(paste(\"Removing genes:\", paste(names(datExpr)[!gsg$goodGenes], collapse = \", \")))\n",
        "    if (sum(!gsg$goodSamples) > 0)\n",
        "        printFlush(paste(\"Removing samples:\", paste(rownames(datExpr)[!gsg$goodSamples], collapse = \", \")))\n",
        "    datExpr <- datExpr[gsg$goodSamples, gsg$goodGenes]\n",
        "}\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1sTu8dU3TpS1",
        "outputId": "36e33f93-624e-400d-ba5b-4269148110df"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " Flagging genes and samples with too many missing values...\n",
            "  ..step 1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Sample Clustering to Detect Outliers"
      ],
      "metadata": {
        "id": "VSnWSskTT5Xy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Cluster samples to see if there are obvious outliers\n",
        "sampleTree <- hclust(dist(datExpr), method = \"average\")\n",
        "\n",
        "# Plot sample clustering\n",
        "pdf(\"1_sample_clustering.pdf\", width = 12, height = 9)\n",
        "par(cex = 0.6)\n",
        "par(mar = c(0,4,2,0))\n",
        "plot(sampleTree, main = \"Sample Clustering to Detect Outliers\",\n",
        "     sub = \"\", xlab = \"\", cex.lab = 1.5, cex.axis = 1.5, cex.main = 2)\n",
        "dev.off()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "id": "D1BsCgMrzYdX",
        "outputId": "651de2c7-ab92-41a3-d8ad-58fe4a13eede"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<strong>agg_record_4601767:</strong> 2"
            ],
            "text/markdown": "**agg_record_4601767:** 2",
            "text/latex": "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_4601767:} 2",
            "text/plain": [
              "agg_record_4601767 \n",
              "                 2 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Preparing Trait Data"
      ],
      "metadata": {
        "id": "ODbQ4EcizeDJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Convert conditions to binary trait matrix for correlation analysis\n",
        "# Create binary variables for each condition\n",
        "trait_data <- data.frame(\n",
        "    Normal = as.numeric(col_data$Conditions == \"Normal Human Breast Organoids\"),\n",
        "    HER2 = as.numeric(col_data$Conditions == \"HER2 Positive Breast Tumor\"),\n",
        "    NonTNBC = as.numeric(col_data$Conditions == \"Non-TNBC Breast Tumor\"),\n",
        "    TNBC = as.numeric(col_data$Conditions == \"TNBC Breast Tumor\"),\n",
        "    row.names = col_data$Sample\n",
        ")\n",
        "\n",
        "# Ensure trait data matches datExpr sample order\n",
        "trait_data <- trait_data[rownames(datExpr), ]\n",
        "\n"
      ],
      "metadata": {
        "id": "S7Td6UhJzeW3"
      },
      "execution_count": 29,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Choosing Soft-Thresholding Power (β)"
      ],
      "metadata": {
        "id": "lyxI9b8RxIs4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Choosing a set of soft-thresholding powers\n",
        "powers <- c(c(1:10), seq(from = 12, to = 20, by = 2))\n",
        "\n",
        "# Call the network topology analysis function\n",
        "sft <- pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)\n",
        "\n",
        "# Plot the results\n",
        "pdf(\"2_scale_independence.pdf\", width = 12, height = 9)\n",
        "par(mfrow = c(1,2))\n",
        "cex1 = 0.9\n",
        "\n",
        "# Scale-free topology fit index as a function of the soft-thresholding power\n",
        "plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],\n",
        "     xlab = \"Soft Threshold (power)\",\n",
        "     ylab = \"Scale Free Topology Model Fit, signed R^2\",\n",
        "     type = \"n\", main = paste(\"Scale Independence\"))\n",
        "text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],\n",
        "     labels = powers, cex = cex1, col = \"red\")\n",
        "abline(h = 0.80, col = \"red\")  # Threshold line at R^2 = 0.80\n",
        "\n",
        "# Mean connectivity as a function of the soft-thresholding power\n",
        "plot(sft$fitIndices[,1], sft$fitIndices[,5],\n",
        "     xlab = \"Soft Threshold (power)\", ylab = \"Mean Connectivity\",\n",
        "     type = \"n\", main = paste(\"Mean Connectivity\"))\n",
        "text(sft$fitIndices[,1], sft$fitIndices[,5],\n",
        "     labels = powers, cex = cex1, col = \"red\")\n",
        "dev.off()\n",
        "\n",
        "# Choose power based on scale-free topology criterion\n",
        "# Typically choose the lowest power where R^2 > 0.80\n",
        "softPower <- sft$powerEstimate\n",
        "if (is.na(softPower)) {\n",
        "    # If automatic selection fails, choose manually (e.g., 6 or 8)\n",
        "    softPower <- 6\n",
        "    cat(\"Automatic power selection failed. Using softPower =\", softPower, \"\\n\")\n",
        "} else {\n",
        "    cat(\"Chosen soft-thresholding power:\", softPower, \"\\n\")\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 403
        },
        "id": "NI6Z3ZSrUIFH",
        "outputId": "5595e797-4859-44bc-cc23-bf4dd9679d0d"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "pickSoftThreshold: will use block size 4049.\n",
            " pickSoftThreshold: calculating connectivity for given powers...\n",
            "   ..working on genes 1 through 4049 of 4049\n",
            "   Power SFT.R.sq   slope truncated.R.sq mean.k. median.k. max.k.\n",
            "1      1   0.4850  2.3100          0.809  1390.0    1450.0   1970\n",
            "2      2   0.2680  0.6190          0.753   711.0     723.0   1250\n",
            "3      3   0.0087 -0.0497          0.606   451.0     419.0    936\n",
            "4      4   0.5180 -0.4440          0.653   322.0     272.0    784\n",
            "5      5   0.6940 -0.6600          0.726   246.0     189.0    686\n",
            "6      6   0.7480 -0.7820          0.768   197.0     137.0    613\n",
            "7      7   0.7730 -0.8640          0.790   162.0     105.0    554\n",
            "8      8   0.7760 -0.9340          0.794   136.0      82.7    504\n",
            "9      9   0.7760 -0.9770          0.808   116.0      66.6    462\n",
            "10    10   0.7830 -1.0200          0.819   100.0      54.4    425\n",
            "11    12   0.7650 -1.0900          0.813    76.8      36.9    363\n",
            "12    14   0.7740 -1.1300          0.840    60.6      25.6    314\n",
            "13    16   0.7890 -1.1600          0.864    48.8      18.8    274\n",
            "14    18   0.7890 -1.1900          0.869    40.1      13.7    241\n",
            "15    20   0.7890 -1.2100          0.875    33.4      10.3    214\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<strong>agg_record_1821400915:</strong> 2"
            ],
            "text/markdown": "**agg_record_1821400915:** 2",
            "text/latex": "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_1821400915:} 2",
            "text/plain": [
              "agg_record_1821400915 \n",
              "                    2 "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Automatic power selection failed. Using softPower = 6 \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Construct Network and Identify Modules"
      ],
      "metadata": {
        "id": "Ap_V3ucgUR8O"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "net <- blockwiseModules(\n",
        "  datExpr,\n",
        "  power = softPower,\n",
        "  corType = \"bicor\",\n",
        "  maxPOutliers = 0.1,\n",
        "  TOMType = \"unsigned\",\n",
        "  minModuleSize = 30,\n",
        "  mergeCutHeight = 0.25,\n",
        "  numericLabels = TRUE,\n",
        "  pamRespectsDendro = FALSE,\n",
        "  saveTOMs = TRUE,\n",
        "  verbose = 3\n",
        ")\n",
        "\n",
        "# Module assignments\n",
        "moduleLabels <- net$colors\n",
        "moduleColors <- labels2colors(net$colors)\n",
        "table(moduleColors)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "UJWQgcADzi9w",
        "outputId": "c17e7e6b-d0c1-4488-8ca1-779f4e8e980e"
      },
      "execution_count": 31,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " Calculating module eigengenes block-wise from all genes\n",
            "   Flagging genes and samples with too many missing values...\n",
            "    ..step 1\n",
            " ..Working on block 1 .\n",
            "    TOM calculation: adjacency..\n",
            "    ..will use 2 parallel threads.\n",
            "     Fraction of slow calculations: 0.000000\n",
            "    ..connectivity..\n",
            "    ..matrix multiplication (system BLAS)..\n",
            "    ..normalization..\n",
            "    ..done.\n",
            "   ..saving TOM for block 1 into file blockwiseTOM-block.1.RData\n",
            " ....clustering..\n",
            " ....detecting modules..\n",
            " ....calculating module eigengenes..\n",
            " ....checking kME in modules..\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(0, 0, 0, 0, 1, 0, 0, 0, 2, 3, 4, 4, 1, 1, 7, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 2 genes from module 1 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(10, 1, 2, 1, 1, 0, 0, 1, 4, 3, 3, 20.04, 6, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n",
            "Warning message in bicor(structure(c(165, 118, 181, 89.04, 21, 35.03, 45, 34.03, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n",
            "Warning message in bicor(structure(c(267.32, 189.3, 229.84, 29.67, 19, 38.51, 10.48, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 2 genes from module 4 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(500.11, 663.85, 905.08, 1406.61, 234.81, 1615.44, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 1 genes from module 5 because their KME is too low.\n",
            "     ..removing 1 genes from module 6 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(177, 271, 222, 140, 96, 244, 176, 266, 495, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n",
            "Warning message in bicor(structure(c(2551, 4061.19, 4622.48, 2794, 962, 1696, 1293, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n",
            "Warning message in bicor(structure(c(180, 409, 157, 212, 334, 574, 1055, 70, 71, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 2 genes from module 10 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(7, 0, 9, 0, 2, 0, 2, 2, 9, 12, 13, 11, 13, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 5 genes from module 11 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(1, 1, 0, 12, 0, 28, 59, 7, 2, 14, 5, 10, 1, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n",
            "Warning message in bicor(structure(c(92.85, 132.39, 100.79, 565.69, 137.22, 266.74, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 1 genes from module 13 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(3, 10, 7, 0, 1, 10, 35, 2, 1, 3, 1, 2, 6, 3, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 1 genes from module 14 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(3565, 3387, 4952, 210, 209, 196, 270, 177, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n",
            "Warning message in bicor(structure(c(25, 16, 29, 1, 4, 10, 5, 3, 10, 26, 20, 40, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 1 genes from module 19 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in bicor(structure(c(96, 49, 54, 21, 74, 801, 157, 232, 92, 22, :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     ..removing 2 genes from module 28 because their KME is too low.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in (function (x, y = NULL, robustX = TRUE, robustY = TRUE, use = \"all.obs\", :\n",
            "“bicor: zero MAD in variable 'x'. Pearson correlation was used for individual columns with zero (or missing) MAD.”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  ..reassigning 1 genes from module 11 to modules with higher KME.\n",
            "  ..reassigning 2 genes from module 13 to modules with higher KME.\n",
            " ..merging modules that are too close..\n",
            "     mergeCloseModules: Merging modules whose distance is less than 0.25\n",
            "       Calculating new MEs...\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "moduleColors\n",
              "     blue     brown     green      grey       red turquoise    yellow \n",
              "      957       783       102        18        31      1484       674 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "yDyLavkoUbt3"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "o43LPYdbud_I"
      },
      "execution_count": 32,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "MEs <- orderMEs(net$MEs)\n",
        "\n",
        "moduleTraitCor <- cor(MEs, trait_data, use = \"p\")\n",
        "moduleTraitPvalue <- corPvalueStudent(moduleTraitCor, nSamples = nrow(datExpr))\n",
        "\n",
        "sig_threshold <- 0.05\n",
        "get_sig_modules <- function(trait) {\n",
        "  p <- moduleTraitPvalue[, trait]\n",
        "  hit <- !is.na(p) & p < sig_threshold\n",
        "  gsub(\"^ME\", \"\", rownames(moduleTraitPvalue)[hit])\n",
        "}\n",
        "\n",
        "her2_modules    <- get_sig_modules(\"HER2\")\n",
        "tnbc_modules    <- get_sig_modules(\"TNBC\")\n",
        "nontnbc_modules <- get_sig_modules(\"NonTNBC\")\n",
        "\n",
        "her2_modules\n",
        "tnbc_modules\n",
        "nontnbc_modules\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 54
        },
        "id": "ipw6bLrqvHgc",
        "outputId": "c1f842e3-33ed-436e-97c7-ddc30c94c001"
      },
      "execution_count": 33,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in cor(MEs, trait_data, use = \"p\"):\n",
            "“the standard deviation is zero”\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [],
            "text/markdown": "",
            "text/latex": "",
            "text/plain": [
              "character(0)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [],
            "text/markdown": "",
            "text/latex": "",
            "text/plain": [
              "character(0)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [],
            "text/markdown": "",
            "text/latex": "",
            "text/plain": [
              "character(0)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# ---- Align metadata samples to expression samples (DO THIS ONCE) ----\n",
        "expr_ids_clean <- sub(\"^X\", \"\", rownames(datExpr))\n",
        "expr_ids_clean <- gsub(\"\\\\.\", \"-\", expr_ids_clean)\n",
        "\n",
        "meta_ids_clean <- sub(\"^X\", \"\", as.character(col_data$Sample))\n",
        "meta_ids_clean <- gsub(\"\\\\.\", \"-\", meta_ids_clean)\n",
        "\n",
        "match_idx <- match(expr_ids_clean, meta_ids_clean)\n",
        "stopifnot(sum(!is.na(match_idx)) == nrow(datExpr))\n",
        "\n",
        "col_data2 <- col_data[match_idx, , drop = FALSE]\n",
        "rownames(col_data2) <- rownames(datExpr)\n",
        "\n",
        "# ---- Build trait matrix using *aligned* metadata ----\n",
        "trait_data <- data.frame(\n",
        "  HER2    = as.numeric(col_data2$Conditions == \"HER2 Positive Breast Tumor\"),\n",
        "  TNBC    = as.numeric(col_data2$Conditions == \"TNBC Breast Tumor\"),\n",
        "  NonTNBC = as.numeric(col_data2$Conditions == \"Non-TNBC Breast Tumor\"),\n",
        "  Normal  = as.numeric(col_data2$Conditions == \"Normal Human Breast Organoids\"),\n",
        "  row.names = rownames(datExpr)\n",
        ")\n",
        "\n",
        "print(colSums(trait_data))          # should be 5,5,6,3\n",
        "stopifnot(all(colSums(trait_data) > 0))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "G8ZyRhqSudsi",
        "outputId": "e7ccc1d1-b51d-4dce-ad22-90aa128804a3"
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   HER2    TNBC NonTNBC  Normal \n",
            "      5       5       6       3 \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Visualising Gene Dendrogram and Module Colors"
      ],
      "metadata": {
        "id": "fmEFxnJySIug"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Plot the dendrogram and module colors\n",
        "pdf(\"3_gene_dendrogram_modules.pdf\", width = 12, height = 9)\n",
        "plotDendroAndColors(net$dendrograms[[1]], moduleColors[net$blockGenes[[1]]],\n",
        "                    \"Module colors\",\n",
        "                    dendroLabels = FALSE, hang = 0.03,\n",
        "                    addGuide = TRUE, guideHang = 0.05,\n",
        "                    main = \"Gene Dendrogram and Module Colors\")\n",
        "dev.off()\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "id": "M8ufrI2eUcix",
        "outputId": "ddf689ac-b9d1-4e94-d6a7-90eebc7df172"
      },
      "execution_count": 35,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<strong>agg_record_482500704:</strong> 2"
            ],
            "text/markdown": "**agg_record_482500704:** 2",
            "text/latex": "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_482500704:} 2",
            "text/plain": [
              "agg_record_482500704 \n",
              "                   2 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Relating Modules to Traits"
      ],
      "metadata": {
        "id": "3FiRhpCzVp4h"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Calculate module eigengenes (ME)\n",
        "MEs <- moduleEigengenes(datExpr, moduleColors)$eigengenes\n",
        "\n",
        "# Order MEs by hierarchical clustering\n",
        "MEs <- orderMEs(MEs)\n",
        "\n",
        "# Calculate correlation between MEs and traits\n",
        "moduleTraitCor <- cor(MEs, trait_data, use = \"p\")\n",
        "moduleTraitPvalue <- corPvalueStudent(moduleTraitCor, nrow(datExpr))\n",
        "\n",
        "# Save results\n",
        "write.csv(moduleTraitCor, \"module_trait_correlation.csv\")\n",
        "write.csv(moduleTraitPvalue, \"module_trait_pvalue.csv\")\n"
      ],
      "metadata": {
        "id": "k6qf90wXVqMe"
      },
      "execution_count": 36,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Visualize Module-Trait Relationships"
      ],
      "metadata": {
        "id": "9P-kD4muWln1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Create a heatmap of module-trait correlations\n",
        "pdf(\"4_module_trait_heatmap.pdf\", width = 8, height = 10)\n",
        "# Display correlation values and p-values\n",
        "textMatrix <- paste(signif(moduleTraitCor, 2), \"\\n(\",\n",
        "                   signif(moduleTraitPvalue, 1), \")\", sep = \"\")\n",
        "dim(textMatrix) <- dim(moduleTraitCor)\n",
        "\n",
        "par(mar = c(6, 8.5, 3, 3))\n",
        "labeledHeatmap(Matrix = moduleTraitCor,\n",
        "               xLabels = names(trait_data),\n",
        "               yLabels = names(MEs),\n",
        "               ySymbols = names(MEs),\n",
        "               colorLabels = FALSE,\n",
        "               colors = blueWhiteRed(50),\n",
        "               textMatrix = textMatrix,\n",
        "               setStdMargins = FALSE,\n",
        "               cex.text = 0.5,\n",
        "               zlim = c(-1,1),\n",
        "               main = paste(\"Module-Trait Relationships\"))\n",
        "dev.off()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "id": "VI4Xud5sSIEa",
        "outputId": "8ee84cd4-57d3-452d-f1b7-49ad5aed274b"
      },
      "execution_count": 37,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<strong>agg_record_2133273548:</strong> 2"
            ],
            "text/markdown": "**agg_record_2133273548:** 2",
            "text/latex": "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_2133273548:} 2",
            "text/plain": [
              "agg_record_2133273548 \n",
              "                    2 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Identifying Modules of Interest"
      ],
      "metadata": {
        "id": "tx8f_HPLWCdK"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define significance threshold\n",
        "sig_threshold <- 0.05\n",
        "\n",
        "# Find modules significantly correlated with HER2\n",
        "her2_modules <- rownames(moduleTraitPvalue)[moduleTraitPvalue[, \"HER2\"] < sig_threshold]\n",
        "cat(\"\\nModules significantly correlated with HER2:\\n\")\n",
        "print(her2_modules)\n",
        "\n",
        "# Find modules significantly correlated with TNBC\n",
        "tnbc_modules <- rownames(moduleTraitPvalue)[moduleTraitPvalue[, \"TNBC\"] < sig_threshold]\n",
        "cat(\"\\nModules significantly correlated with TNBC:\\n\")\n",
        "print(tnbc_modules)\n",
        "\n",
        "# Find modules significantly correlated with NonTNBC\n",
        "nontnbc_modules <- rownames(moduleTraitPvalue)[moduleTraitPvalue[, \"NonTNBC\"] < sig_threshold]\n",
        "cat(\"\\nModules significantly correlated with NonTNBC:\\n\")\n",
        "print(nontnbc_modules)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wy6rELzgWCyF",
        "outputId": "c8f9c119-0f1f-4899-eb68-8f39787bd1f4"
      },
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Modules significantly correlated with HER2:\n",
            "[1] \"MEyellow\"    \"MEturquoise\" \"MEred\"       \"MEblue\"      \"MEgrey\"     \n",
            "\n",
            "Modules significantly correlated with TNBC:\n",
            "[1] \"MEyellow\"    \"MEturquoise\" \"MEred\"       \"MEbrown\"     \"MEgrey\"     \n",
            "\n",
            "Modules significantly correlated with NonTNBC:\n",
            "[1] \"MEred\"\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "### Exporting Gene Lists for Significant Modules\n"
      ],
      "metadata": {
        "id": "ECLv2orXWXCR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Get gene names for each module\n",
        "genes <- colnames(datExpr)\n",
        "module_gene_list <- data.frame(\n",
        "    gene_id = genes,\n",
        "    module_color = moduleColors,\n",
        "    stringsAsFactors = FALSE\n",
        ")\n",
        "\n",
        "# Save complete module assignments\n",
        "write.csv(module_gene_list, \"module_gene_assignments.csv\", row.names = FALSE)\n",
        "\n",
        "# Export genes for each significant module\n",
        "all_sig_modules <- unique(c(her2_modules, tnbc_modules, nontnbc_modules))\n",
        "\n",
        "for (module in all_sig_modules) {\n",
        "    module_name <- gsub(\"ME\", \"\", module)\n",
        "    module_genes <- genes[moduleColors == module_name]\n",
        "    write.table(module_genes,\n",
        "                file = paste0(\"genes_module_\", module_name, \".txt\"),\n",
        "                row.names = FALSE, col.names = FALSE, quote = FALSE)\n",
        "    cat(\"Module\", module_name, \"contains\", length(module_genes), \"genes\\n\")\n",
        "}\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oZ6034wjWXbY",
        "outputId": "0a62cccb-e894-4c2a-e2de-b2363010fba7"
      },
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Module yellow contains 674 genes\n",
            "Module turquoise contains 1484 genes\n",
            "Module red contains 31 genes\n",
            "Module blue contains 957 genes\n",
            "Module grey contains 18 genes\n",
            "Module brown contains 783 genes\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "### HUB Gene Extraction"
      ],
      "metadata": {
        "id": "Uw1aaIf8YAgc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Collecting ALL significant modules across traits\n",
        "sig_MEs <- unique(c(her2_modules, tnbc_modules, nontnbc_modules))\n",
        "sig_MEs <- unique(na.omit(sig_MEs))\n",
        "\n",
        "# Remove grey if present\n",
        "sig_MEs <- sig_MEs[sig_MEs != \"ME0\"]\n",
        "\n",
        "print(sig_MEs)  # sanity check: should be like \"ME1\" \"ME4\" \"ME6\" ...\n",
        "\n",
        "#kME matrix (genes x modules)\n",
        "kME <- as.data.frame(cor(datExpr, net$MEs, use = \"p\"))\n",
        "colnames(kME) <- colnames(net$MEs)\n",
        "\n",
        "# Mapping numeric labels -> colours (for filenames / readability)\n",
        "label_to_color <- setNames(\n",
        "  WGCNA::labels2colors(sort(unique(moduleLabels))),\n",
        "  sort(unique(moduleLabels))\n",
        ")\n",
        "\n",
        "# Looping through significant modules\n",
        "for (MEname in sig_MEs) {\n",
        "\n",
        "  # Extracting numeric module label (ME1 -> 1)\n",
        "  lab <- suppressWarnings(as.integer(sub(\"^ME\", \"\", MEname)))\n",
        "  if (is.na(lab) || lab == 0) next\n",
        "\n",
        "  inModule <- moduleLabels == lab\n",
        "  if (!any(inModule)) next\n",
        "\n",
        "  # kME for genes in this module\n",
        "  kME_mod <- kME[inModule, MEname, drop = FALSE]\n",
        "  kME_mod <- kME_mod[order(-abs(kME_mod[[MEname]])), , drop = FALSE]\n",
        "\n",
        "  top <- head(kME_mod, 20)\n",
        "\n",
        "  module_color <- label_to_color[as.character(lab)]\n",
        "\n",
        "  write.csv(\n",
        "    data.frame(\n",
        "      gene = rownames(top),\n",
        "      kME  = top[[MEname]]\n",
        "    ),\n",
        "    file = paste0(\"hub_genes_\", MEname, \"_\", module_color, \".csv\"),\n",
        "    row.names = FALSE\n",
        "  )\n",
        "\n",
        "  cat(\"\\nTop 10 hub genes for\", MEname, \"(\", module_color, \"):\\n\")\n",
        "  print(head(rownames(top), 10))\n",
        "}"
      ],
      "metadata": {
        "id": "S1r2x8y8qreT",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d1b76a7a-78b0-45bf-8124-fbe9e3653f14"
      },
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[1] \"MEyellow\"    \"MEturquoise\" \"MEred\"       \"MEblue\"      \"MEgrey\"     \n",
            "[6] \"MEbrown\"    \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Converting Colour-based significant modules to Numeric ME labels"
      ],
      "metadata": {
        "id": "wtyWm5rEA-Wv"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Build colour -> numeric label mapping from the network\n",
        "color_to_label <- tapply(moduleLabels, moduleColors, function(x) {\n",
        "  as.integer(names(sort(table(x), decreasing = TRUE))[1])\n",
        "})\n",
        "\n",
        "cat(\"\\nColour → numeric label mapping:\\n\")\n",
        "print(color_to_label)\n",
        "\n",
        "convert_color_MEs_to_numeric <- function(MEs_color) {\n",
        "  cols <- gsub(\"^ME\", \"\", MEs_color)\n",
        "  labs <- color_to_label[cols]\n",
        "  labs <- labs[!is.na(labs) & labs != 0]   # drop NA and grey\n",
        "  paste0(\"ME\", labs)\n",
        "}\n",
        "\n",
        "# Convert significant module lists\n",
        "her2_modules_num    <- convert_color_MEs_to_numeric(her2_modules)\n",
        "tnbc_modules_num    <- convert_color_MEs_to_numeric(tnbc_modules)\n",
        "nontnbc_modules_num <- convert_color_MEs_to_numeric(nontnbc_modules)\n",
        "\n",
        "# Combine all significant numeric MEs\n",
        "sig_MEs_numeric <- unique(c(her2_modules_num,\n",
        "                            tnbc_modules_num,\n",
        "                            nontnbc_modules_num))\n",
        "\n",
        "cat(\"\\nSignificant numeric ME labels:\\n\")\n",
        "print(sig_MEs_numeric)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fF-Rn2FqCHil",
        "outputId": "377731b1-a4ff-4bbb-8e0c-9b7ea9cb0754"
      },
      "execution_count": 43,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Colour → numeric label mapping:\n",
            "     blue     brown     green      grey       red turquoise    yellow \n",
            "        2         3         5         0         6         1         4 \n",
            "\n",
            "Significant numeric ME labels:\n",
            "[1] \"ME4\" \"ME1\" \"ME6\" \"ME2\" \"ME3\"\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Hub Gene Extraction Using Numeric ME Labels"
      ],
      "metadata": {
        "id": "uKBpxkTtIteg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Use the numeric ME labels produced in the last Step\n",
        "sig_MEs <- sig_MEs_numeric\n",
        "print(sig_MEs)\n",
        "\n",
        "# kME matrix: genes x module eigengenes\n",
        "kME <- as.data.frame(cor(datExpr, net$MEs, use = \"p\"))\n",
        "colnames(kME) <- colnames(net$MEs)\n",
        "\n",
        "# Map numeric labels to colours (for filenames / readability)\n",
        "label_to_color <- setNames(\n",
        "  WGCNA::labels2colors(sort(unique(moduleLabels))),\n",
        "  sort(unique(moduleLabels))\n",
        ")\n",
        "\n",
        "for (MEname in sig_MEs) {\n",
        "\n",
        "  # Extract numeric label (ME4 -> 4)\n",
        "  lab <- suppressWarnings(as.integer(sub(\"^ME\", \"\", MEname)))\n",
        "  if (is.na(lab) || lab == 0) next\n",
        "\n",
        "  inModule <- moduleLabels == lab\n",
        "  if (!any(inModule)) next\n",
        "\n",
        "  # kME for genes in this module\n",
        "  kME_mod <- kME[inModule, MEname, drop = FALSE]\n",
        "  kME_mod <- kME_mod[order(-abs(kME_mod[[MEname]])), , drop = FALSE]\n",
        "\n",
        "  top <- head(kME_mod, 20)\n",
        "\n",
        "  module_color <- label_to_color[as.character(lab)]\n",
        "\n",
        "  out_file <- paste0(\"hub_genes_\", MEname, \"_\", module_color, \".csv\")\n",
        "\n",
        "  write.csv(\n",
        "    data.frame(\n",
        "      gene = rownames(top),\n",
        "      kME  = top[[MEname]]\n",
        "    ),\n",
        "    file = out_file,\n",
        "    row.names = FALSE\n",
        "  )\n",
        "\n",
        "  cat(\"\\nTop 10 hub genes for\", MEname, \"(\", module_color, \"):\\n\", sep = \"\")\n",
        "  print(head(rownames(top), 10))\n",
        "  cat(\"Saved:\", out_file, \"\\n\")\n",
        "}\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HCJz_i2oCHXx",
        "outputId": "0fd1f391-14a4-4cfe-fcf1-bbef7d953e80"
      },
      "execution_count": 45,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[1] \"ME4\" \"ME1\" \"ME6\" \"ME2\" \"ME3\"\n",
            "\n",
            "Top 10 hub genes forME4(yellow):\n",
            " [1] \"ENSG00000229794\" \"ENSG00000176124\" \"ENSG00000270314\" \"ENSG00000206634\"\n",
            " [5] \"ENSG00000206838\" \"ENSG00000225572\" \"ENSG00000206596\" \"ENSG00000207005\"\n",
            " [9] \"ENSG00000207389\" \"ENSG00000273768\"\n",
            "Saved: hub_genes_ME4_yellow.csv \n",
            "\n",
            "Top 10 hub genes forME1(turquoise):\n",
            " [1] \"ENSG00000149452\" \"ENSG00000172519\" \"ENSG00000205277\" \"ENSG00000226741\"\n",
            " [5] \"ENSG00000168367\" \"ENSG00000260902\" \"ENSG00000177992\" \"ENSG00000287514\"\n",
            " [9] \"ENSG00000242512\" \"ENSG00000169876\"\n",
            "Saved: hub_genes_ME1_turquoise.csv \n",
            "\n",
            "Top 10 hub genes forME6(red):\n",
            " [1] \"ENSG00000065361\" \"ENSG00000178568\" \"ENSG00000084628\" \"ENSG00000170871\"\n",
            " [5] \"ENSG00000111058\" \"ENSG00000291083\" \"ENSG00000142973\" \"ENSG00000168004\"\n",
            " [9] \"ENSG00000186665\" \"ENSG00000135622\"\n",
            "Saved: hub_genes_ME6_red.csv \n",
            "\n",
            "Top 10 hub genes forME2(blue):\n",
            " [1] \"ENSG00000137218\" \"ENSG00000123240\" \"ENSG00000181027\" \"ENSG00000121775\"\n",
            " [5] \"ENSG00000161265\" \"ENSG00000100242\" \"ENSG00000128563\" \"ENSG00000123144\"\n",
            " [9] \"ENSG00000142409\" \"ENSG00000230979\"\n",
            "Saved: hub_genes_ME2_blue.csv \n",
            "\n",
            "Top 10 hub genes forME3(brown):\n",
            " [1] \"ENSG00000083535\" \"ENSG00000164329\" \"ENSG00000144028\" \"ENSG00000128908\"\n",
            " [5] \"ENSG00000168826\" \"ENSG00000111596\" \"ENSG00000010244\" \"ENSG00000132842\"\n",
            " [9] \"ENSG00000082258\" \"ENSG00000130939\"\n",
            "Saved: hub_genes_ME3_brown.csv \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### # Reassign canonical module lists to numeric ME labels"
      ],
      "metadata": {
        "id": "2HhPupFiDtGs"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "her2_modules    <- her2_MEs_numeric\n",
        "tnbc_modules    <- tnbc_MEs_numeric\n",
        "nontnbc_modules <- nontnbc_MEs_numeric\n",
        "\n",
        "# Sanity check\n",
        "cat(\"HER2 numeric modules:\\n\")\n",
        "print(her2_modules)\n",
        "\n",
        "cat(\"TNBC numeric modules:\\n\")\n",
        "print(tnbc_modules)\n",
        "\n",
        "cat(\"NonTNBC numeric modules:\\n\")\n",
        "print(nontnbc_modules)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YSqdUUA-DtY3",
        "outputId": "9fa3f1ce-3415-4ff9-a2bc-53db72d15e3b"
      },
      "execution_count": 47,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "HER2 numeric modules:\n",
            "[1] \"ME4\" \"ME1\" \"ME6\" \"ME2\"\n",
            "TNBC numeric modules:\n",
            "[1] \"ME4\" \"ME1\" \"ME6\" \"ME3\"\n",
            "NonTNBC numeric modules:\n",
            "[1] \"ME6\"\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Summary Statistics"
      ],
      "metadata": {
        "id": "Y19apbBpYxaJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# ===== FINAL WGCNA SUMMARY (REPORTING ONLY) =====\n",
        "\n",
        "label_to_color <- setNames(\n",
        "  WGCNA::labels2colors(sort(unique(moduleLabels))),\n",
        "  sort(unique(moduleLabels))\n",
        ")\n",
        "\n",
        "clean_modules <- function(x) {\n",
        "  x <- unique(na.omit(x))\n",
        "  x <- sub(\"^ME\", \"\", x)\n",
        "  x <- x[x != \"0\"]\n",
        "  x <- suppressWarnings(as.integer(x))\n",
        "  x <- x[!is.na(x)]\n",
        "  if (length(x) == 0) \"None\" else paste(label_to_color[x], collapse = \", \")\n",
        "}\n",
        "\n",
        "cat(\"Total genes analyzed:\", ncol(datExpr), \"\\n\")\n",
        "cat(\"Total samples:\", nrow(datExpr), \"\\n\")\n",
        "cat(\"Soft-thresholding power (β):\", softPower, \"\\n\")\n",
        "cat(\"Number of modules identified (excluding grey):\",\n",
        "    length(setdiff(unique(moduleLabels), 0)), \"\\n\")\n",
        "\n",
        "cat(\"\\nModule sizes:\\n\")\n",
        "print(table(moduleColors))\n",
        "\n",
        "cat(\"\\nModules significantly associated with traits (p < 0.05):\\n\")\n",
        "cat(\"  HER2:\", clean_modules(her2_modules), \"\\n\")\n",
        "cat(\"  TNBC:\", clean_modules(tnbc_modules), \"\\n\")\n",
        "cat(\"  NonTNBC:\", clean_modules(nontnbc_modules), \"\\n\")\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hwGVv3G5xJCa",
        "outputId": "9a2f603f-5b0f-4b97-a079-6382b4f2c2e6"
      },
      "execution_count": 51,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Total genes analyzed: 4049 \n",
            "Total samples: 19 \n",
            "Soft-thresholding power (β): 6 \n",
            "Number of modules identified (excluding grey): 6 \n",
            "\n",
            "Module sizes:\n",
            "moduleColors\n",
            "     blue     brown     green      grey       red turquoise    yellow \n",
            "      957       783       102        18        31      1484       674 \n",
            "\n",
            "Modules significantly associated with traits (p < 0.05):\n",
            "  HER2: brown, grey, green, turquoise \n",
            "  TNBC: brown, grey, green, blue \n",
            "  NonTNBC: green \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Saving R workspace for future use"
      ],
      "metadata": {
        "id": "ZwPC0z0jY2kZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "save.image(\"WGCNA_analysis_final.RData\")"
      ],
      "metadata": {
        "id": "cfq5u-QZY28X"
      },
      "execution_count": 49,
      "outputs": []
    }
  ]
}